### PR TITLE
chore: move the authentication middleware into droplet framework 

### LINKED ISSUE #1295: Chore: move the authentication middleware into droplet framework
# Feature request

## Please describe your feature

Hi, I found the [dorplet](https://github.com/ShiningRush/droplet/tree/master/middleware) had split the http_input into `http_injector` and `http_input`. We can get the http.Request in droplect.Context.
So I think we can move those middleware:

https://github.com/apache/apisix-dashboard/blob/822aa6d1f40c5eb5bfef00030304eb5b619a9eea/api/internal/route.go#L56-L58

into droplet framework. Then we can have a more clearer code and better response.

First, I want to move the authentication function into droplet.

## Describe the solution you'd like

Use the droplet framework to do middleware orchestra work.

https://github.com/apache/apisix-dashboard/blob/5f2232603588a47f902a90d974692a7c6bea535f/api/cmd/managerapi.go#L61-L68

### PR DESCRIPTION
**Pull request by starsz**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [ ] New feature provided
- [x] Improve performance
- [ ] Backport patches

- Related issues
fix #1295
___
### New feature or improvement
- Describe the details and related test reports.

chore: move the authentication middleware into droplet framework. And add the middleware unit test.
___
### Please add the corresponding test cases if necessary.

Added.

### CODE REVIEW

File: api/cmd/managerapi.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ package cmd
 import (
 	"context"
 	"fmt"
+	"github.com/apisix/manager-api/internal/filter"
```
Comment (nic-chen) [inline api/cmd/managerapi.go:22]: code style

File: api/go.mod [hunk 2]
```diff
@@ -21,6 +21,7 @@ require (
 	github.com/gin-contrib/sessions v0.0.3
 	github.com/gin-contrib/static v0.0.0-20200916080430-d45d9a37d28e
 	github.com/gin-gonic/gin v1.6.3
+	github.com/go-playground/assert/v2 v2.0.1
```
Comment (nic-chen) [inline api/go.mod:24]: we have another `assert` lib, I think we should use the same one.

File: api/cmd/managerapi.go [hunk 3]
```diff
@@ -62,7 +63,7 @@ func NewManagerAPICommand() *cobra.Command {
 				var newMws []droplet.Middleware
 				// default middleware order: resp_reshape, auto_input, traffic_log
 				// We should put err_transform at second to catch all error
-				newMws = append(newMws, mws[0], &handler.ErrorTransformMiddleware{})
+				newMws = append(newMws, mws[0], &handler.ErrorTransformMiddleware{}, &filter.AuthenticationMiddleware{})
```
Comment (nic-chen) [inline api/cmd/managerapi.go:66]: just confirm that will other filters be executed if user is not login?

Comment (starsz) [reply api/cmd/managerapi.go:66]: Yes. It will.
Those middlewares are in the gin's API handler.

File: api/go.mod (hunk 2 above)
Comment (starsz) [reply api/go.mod:24]: OK.Let me fix it.

File: api/cmd/managerapi.go (hunk 1 above)
Comment (starsz) [reply api/cmd/managerapi.go]: fixed.

**Review by nic-chen — APPROVED** [review]

File: api/internal/filter/authentication.go [hunk 4]
```diff
@@ -17,62 +17,79 @@
...
-			}
+type AuthenticationMiddleware struct {
+	middleware.BaseMiddleware
+}
+
+func (mw *AuthenticationMiddleware) Handle(ctx droplet.Context) error {
+	httpReq := ctx.Get(middleware.KeyHttpRequest)
+	if httpReq == nil {
+		err := errors.New("input middleware cannot get http request, " +
+			"please check if HttpInfoInjectorMiddleware middlle work well")
```
Comment (tokers) [inline api/internal/filter/authentication.go:41]: For me, it's more like a "program fault" not an exception, i'm prefer to panic here.

Comment (starsz) [reply api/internal/filter/authentication.go:41]: OK.Let me change it.

Comment (starsz) [reply api/internal/filter/authentication.go:41]: OK, let me change it.

**Review by juzhiyuan — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Pull request merged**